src/diff_delta.h

```diff
@@ -0,0 +1,321 @@
+// Copyright (c) 2015 The Guldencoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+//
+// This file contains Delta, the Guldencoin Difficulty Re-adjustment algorithm developed by Frank (dt_cdog@yahoo.com) with various enhancements by Malcolm MacLeod (mmacleod@webmail.co.za)
+// The core algorithm works by taking time measurements of four periods (last block; short window; medium window; long window) and then apply a weighting to them.
+// This allows the algorithm to react to short term fluctuations while still taking long term block targets into account, which helps prevent it from overreacting.
+//
+// In addition to the core algorithm several extra rules are then applied in certain situations (e.g. multiple quick blocks) to enhance the behaviour.
+
+
+#ifndef GULDENCOIN_DIFF_DELTA_H
+#define GULDENCOIN_DIFF_DELTA_H
+
+#define PERCENT_FACTOR 100
+unsigned int static GetNextWorkRequired_Delta(const CBlockIndex* pindexLast, const CBlockHeader *pblock, unsigned int nFirstDeltaBlock)
+{
+    // These two variables are not used in the calculation at all, but only for logging when -debug is set, to prevent logging the same calculation repeatedly.
+    static int64_t nPrevHeight     = 0;
+    static int64_t nPrevDifficulty = 0;
+    
+    std::string sLogInfo;
+
+    // The spacing we want our blocks to come in at.
+    int64_t nRetargetTimespan      = nTargetSpacing;
+
+    // The minimum difficulty that is allowed, this is set on a per algorithm basis.
+    const unsigned int nProofOfWorkLimit = bnProofOfWorkLimit.GetCompact();
+
+    // How many blocks to use to calculate each of the four algo specific time windows (last block; short window; middle window; long window)
+    const int nLastBlock           =   1;
+    const int nShortFrame          =   3;
+    const int nMiddleFrame         =  24;
+    const int nLongFrame           = 576;
+
+    // How many blocks to use for the fifth window, fifth window is across all algorithms.
+    #ifdef MULTI_ALGO
+    const int nDayFrame            = 576;
+    #endif
+
+    // Weighting to use for each of the four algo specific time windows.
+    const int64_t nLBWeight        =  64;
+    const int64_t nShortWeight     =  8;
+    int64_t nMiddleWeight          =  2;
+    int64_t nLongWeight            =  1;
```

Comment:
There is a special case (if more than four blocks come in faster than X
and the latest block is 'under time') where we ignore the middle and
long time frames entirely - these variables need to be set to 0 for that
case.


> In src/diff_delta.h:
> 
> > +    // How many blocks to use to calculate each of the four algo specific time windows (last block; short window; middle window; long window)
> > +    const int nLastBlock           =   1;
> > +    const int nShortFrame          =   3;
> > +    const int nMiddleFrame         =  24;
> > +    const int nLongFrame           = 576;
> > +
> > +    // How many blocks to use for the fifth window, fifth window is across all algorithms.
> > +    #ifdef MULTI_ALGO
> > +    const int nDayFrame            = 576;
> > +    #endif
> > +
> > +    // Weighting to use for each of the four algo specific time windows.
> > +    const int64_t nLBWeight        =  64;
> > +    const int64_t nShortWeight     =  8;
> > +    int64_t nMiddleWeight          =  2;
> > +    int64_t nLongWeight            =  1;
> 
> Just curious, why are these not const?
> 
> —
> Reply to this email directly or view it on GitHub.
> 
>